pkg/capacity/controller.go

```diff
@@ -0,0 +1,562 @@
+		topologyInformer: topologyInformer,
+		scInformer:       scInformer,
+		cInformer:        cInformer,
+		pollPeriod:       pollPeriod,
+		capacities:       map[workItem]*storagev1alpha1.CSIStorageCapacity{},
+	}
+
+	// Now register for changes. Depending on the implementation of the informers,
+	// this may already invoke callbacks.
+	handler := cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.onSCAddOrUpdate(obj.(*storagev1.StorageClass)) },
+		UpdateFunc: func(_ interface{}, newObj interface{}) { c.onSCAddOrUpdate(newObj.(*storagev1.StorageClass)) },
+		DeleteFunc: func(obj interface{}) { c.onSCDelete(obj.(*storagev1.StorageClass)) },
+	}
+	c.scInformer.Informer().AddEventHandler(handler)
+	c.topologyInformer.AddCallback(c.onTopologyChanges)
+
+	// We don't want the callbacks yet, but need to ensure that
+	// the informer controller is instantiated before the caller
+	// starts the factory.
+	cInformer.Informer()
+
+	return c
+}
+
+// Run is a main Controller handler
+func (c *Controller) Run(ctx context.Context, threadiness int) {
+	klog.Info("Starting Capacity Controller")
+	defer c.queue.ShutDown()
+	go c.scInformer.Informer().Run(ctx.Done())
+	go c.topologyInformer.Run(ctx)
+
+	c.prepare(ctx)
+	for i := 0; i < threadiness; i++ {
+		go wait.UntilWithContext(ctx, func(ctx context.Context) {
+			c.runWorker(ctx)
+		}, time.Second)
+	}
+
+	go wait.UntilWithContext(ctx, func(ctx context.Context) { c.pollCapacities() }, c.pollPeriod)
+
+	klog.Info("Started Capacity Controller")
+	<-ctx.Done()
+	klog.Info("Shutting down Capacity Controller")
+}
+
+func (c *Controller) prepare(ctx context.Context) {
+	// Wait for topology and storage class informer sync. Once we have that,
+	// we know which CSIStorageCapacity objects we need.
+	if !cache.WaitForCacheSync(ctx.Done(), c.topologyInformer.HasSynced, c.scInformer.Informer().HasSynced, c.cInformer.Informer().HasSynced) {
+		return
+	}
+	klog.V(3).Infof("Initial number of topology segments %d, storage classes %d, potential CSIStorageCapacity objects %d",
+		len(c.topologyInformer.List()),
+		func() int {
+			scs, _ := c.scInformer.Lister().List(labels.Everything())
+			return len(scs)
+		}(),
+		len(c.capacities))
+
+	// Now that we know what we need, we can check what we have.
+	// We do that both via callbacks *and* by iterating over all
+	// objects: callbacks handle future updates and iterating
+	// avoids the assumumption that the callback will be invoked
+	// for all objects immediately when adding it.
+	klog.V(3).Info("Checking for existing CSIStorageCapacity objects")
+	handler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) { c.onCAddOrUpdate(ctx, obj.(*storagev1alpha1.CSIStorageCapacity)) },
+		UpdateFunc: func(_ interface{}, newObj interface{}) {
+			c.onCAddOrUpdate(ctx, newObj.(*storagev1alpha1.CSIStorageCapacity))
+		},
+		DeleteFunc: func(obj interface{}) { c.onCDelete(ctx, obj.(*storagev1alpha1.CSIStorageCapacity)) },
```

Comment:
Can that ever happen? I've seen the checks elsewhere but didn't add them myself because the callbacks should never get any other kind of object.

Reply:
At least DeleteFunc used to be called with deletion tombstone - not the latest version of the API object, but the latest known by the informer.

https://github.com/kubernetes/kubernetes/blob/ff3e5e06a79bc69ad3d7ccedd277542b6712514b/pkg/controller/volume/persistentvolume/pv_controller_base.go#L180

Reply:
Interesting, I learned something new, thanks. I wonder whether I could have learned that from reading some API description.

I added cast checking. I see that as something that needs to be handled in the glue code rather than the actual callback, but it's now a bit repetitive, so I could be convinced to move casting into the callback...